### PR TITLE
[K8s] Zero config networking for Kubernetes

### PR DESCRIPTION
**Pull request by romilbhardwaj**
This PR introduces new networking features for our Kubernetes support. In particular, we no longer need opening many ports on the Kubernetes cluster nodes. Now we support two modes of operation:

1. `portforward`: Open no ports, and we use `kubectl port-forward` under the hood to reach the pods. This requires zero configuration on the user's end, and is only marginally worse (~10%) in performance (see benchmarks). Given the significantly better UX, this will the default mode of operation. 
2. `nodeport`: Open 1 port, and we run a ssh jump pod on that port to reach other pods. This requires opening one port on any one node in the Kubernetes cluster, and offers the highest performance while minimizing the number of open ports needed.

Users who don't want to use `portforward` can switch to `nodeport` by modifying their `~/.sky/config` file:
```
kubernetes:
  networking: nodeport
```

Note that we currently create one jump pod per user. Eventually, we want to share the jump pod across many users (See #2499)

--------
This PR also has other bug fixes, including populating k8s envvars when the user runs SSH (#2287 and #2453 will also be closed by this PR).

Thanks to @landscapepainter, @aviweit and @hemildesai for their contributions.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"`
- [x] Ensure #2453 is fixed

### CODE REVIEW

**Comment by romilbhardwaj** [comment]
This is ready for review!

File: sky/skylet/providers/kubernetes/node_provider.py
Comment (landscapepainter) [inline sky/skylet/providers/kubernetes/node_provider.py:271]:
```suggestion
            ('printenv | awk -F "=" \'{print "export " $1 "=\\047" $2 "\\047"}\' > ~/k8s_env_var.sh && '
```

**Comment by romilbhardwaj** [comment]
Blocked on #2556. This will likely need minor changes after it is merged. Rest can still be reviewed.

File: sky/utils/kubernetes_utils.py
Comment (Michaelvll) [inline sky/utils/kubernetes_utils.py:942]: Can we add the method for installing `socat` on mac, as `socat` seems required for local machine?
```
brew install socat
```

File: sky/utils/kubernetes_utils.py
Comment (Michaelvll) [inline sky/utils/kubernetes_utils.py:744]: Should we use `debug` instead. It seems the user do not need to know the underlying implementation details.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the PR @romilbhardwaj! Just tested with a newly launched GKE cluster (1 t4, 2 n2-highmem-8) without any network configuration.
Tried the following commands and it works like magic:
```
sky launch -c test-k8s --memory 60+ echo hi
sky launch -c test-k8s-2 --memory 60+ echo hi
sky launch -c test-k8s-3 --gpus t4 nvidia-smi
ssh test-k8s-3; nvidia-smi
```

File: sky/backends/cloud_vm_ray_backend.py
Comment (romilbhardwaj) [reply sky/backends/cloud_vm_ray_backend.py:2337]: No, NodePort would still work - it's just that now everything goes through a SSH Jump Pod, so the SSH port remains fixed at 22 (though the jump port changes in and is fetched dynamically in `kubernetes_utils.get_ssh_proxy_command`).

File: sky/skylet/providers/kubernetes/node_provider.py
Comment (romilbhardwaj) [reply sky/skylet/providers/kubernetes/node_provider.py:322]: Thanks! Reworded now.

File: sky/utils/kubernetes_utils.py
Comment (romilbhardwaj) [reply sky/utils/kubernetes_utils.py:41]: Yes, it's supported. In `kubernetes-ray.yml.j2` we changed the service type to a regular `ClusterIP` instead of `NodePort` since all SSH connections go through a jump ssh pod now. 

The SSH Jump Pod [service can still be NodePort](https://github.com/skypilot-org/skypilot/blob/d8302f0dce4a8f9ee9afadba9c9c7368441ba781/sky/templates/kubernetes-sshjump.yml.j2#L54).

File: sky/utils/kubernetes_utils.py
Comment (romilbhardwaj) [reply sky/utils/kubernetes_utils.py:744]: Agreed! Changed now.

**Comment by romilbhardwaj** [comment]
Thanks for the reviews @Michaelvll! This is ready for another look. 

Running smoke tests on GKE now:
- [ ] `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"` with default port-forward mode
- [ ] `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"` with `nodeport` mode set in ~/.sky/config

> One question I have is whether we would like to preserve the old NodePort way, as it seems we have removed some NodePort related code, not sure if it will still work. Also, for code simplicity, it would be nice if we can remove the old mode, if there is no strong need for it. ; )

That's a good point. The NodePort method is preserved for now since the port-forward mode might be considered as a hack by some (since it relies on tunneling over the API server, and that tunnel is designed only for development work). I was thinking we could collect feedback from users and deprecate it in the future if `port-forward` works fine. In the meanwhile, we have an easy way to switch between methods if `port-forward` doesn't work for them. We are also not documenting the `NodePort` ability for now to make sure users do not use it, unless they really need to.

**Review by Michaelvll — APPROVED** [review]
Thanks for the quick fix @romilbhardwaj! The code looks pretty good to me.

**Comment by romilbhardwaj** [comment]
Thanks for the fast reviews @Michaelvll! Waiting on nodeport smoke tests to pass, will merge after that.

**Pull request merged**
